pkg/provisioner/controller.go

```diff
@@ -291,6 +293,103 @@ func (c *obcController) handleProvisionClaim(key string, obc *v1alpha1.ObjectBuc
 	// to control access to static buckets via RBAC rules on storage classes.
 	isDynamicProvisioning := isNewBucketByStorageClass(class)
 
+	// Re-Get the claim in order to shorten the race condition where the claim was deleted after provisioning started
+	obc, err = claimForKey(key, c.libClientset)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return fmt.Errorf("OBC was lost before we could provision: %v", err)
+		}
+		return err
+	}
+
+	// Older versions of this library set obc.Spec.BucketName with the name of the bucket generated
+	// from obc.Spec.GenerateBucketName, which causes an error condition when updating to a newer
+	// version of this library if provisioning runs again. If the bucket name has the
+	// GenerateBucketName prefix, we can assume that this was originally intended to be a generated
+	// name.
+	if obc.Spec.GenerateBucketName != "" && obc.Spec.BucketName != "" {
+		if strings.HasPrefix(obc.Spec.BucketName, obc.Spec.GenerateBucketName+"-") {
+			obc.Spec.BucketName = ""
+			obc, err = updateClaim(
+				c.libClientset,
+				obc,
+				defaultRetryBaseInterval,
+				defaultRetryTimeout)
+			if err != nil {
+				return fmt.Errorf("error updating OBC to remove set spec.bucketName matching the prefix from spec.generateBucketName: %v", err)
+			}
+		}
+	}
+
+	updateOnSuccess := func() error {
+		// In the case where the OB has been created and the operator crashed immediately after
+		// (before the OB phase could be updated), we should update the OB phase in addition to
+		// updating the OBC reference to the OB and the OBC's phase.
+		ob, err = updateObjectBucketPhase(
+			c.libClientset,
+			ob,
+			v1alpha1.ObjectBucketStatusPhaseBound,
+			defaultRetryBaseInterval,
+			defaultRetryTimeout)
+		if err != nil {
+			return fmt.Errorf("error updating OB %q's status to %q: %v", ob.Name, v1alpha1.ObjectBucketStatusPhaseBound, err)
+		}
+		// update OBC
+		obc.Spec.ObjectBucketName = ob.Name
+		// Do not set obc.Spec.BucketName here: (1) if user has it set, it should be as they set it
+		// (2) if generateBucketName is set, setting this will cause errors if provisioning runs again
+		obc, err = updateClaim(
+			c.libClientset,
+			obc,
+			defaultRetryBaseInterval,
+			defaultRetryTimeout)
+		if err != nil {
+			return fmt.Errorf("error updating OBC: %v", err)
+		}
+		obc, err = updateObjectBucketClaimPhase(
+			c.libClientset,
+			obc,
+			v1alpha1.ObjectBucketClaimStatusPhaseBound,
+			defaultRetryBaseInterval,
+			defaultRetryTimeout)
+		if err != nil {
+			return fmt.Errorf("error updating OBC %q's status to: %v", v1alpha1.ObjectBucketClaimStatusPhaseBound, err)
```

Comment:
does it matter here and the `return` above that the OB has it's Phase set to "Bound"?

Reply:
I'm not quite following the question. I'm sorry.

Reply:
What I noticed is that we 1) update ob's Phase to "Bound", 2) update the obc with the binding OB.name, 3) update the obc's Phase to "Bound".  Does it matter if 1) and 2) or just 1) succeeds but the subsequent updates fail? IOW, assume 1 succeeds and 2 fails: we now have the OB saying bound but the OBC appears not bound. So we return an error and retry which I guess is fine. This comment was more about thinking out loud apparently...

Reply:
I don't think there are any functional issues other than OB displaying "Bound" before OBC reports that. FWIW, existing code in Rook has an infinite OBC provision loop which leaves OB to "Bound" but resets "OBC" to "Pending." Maybe the code that resets "OBC" to "Pending" should also set "OB" to "Pending"?

Reply:
Comment resolved.

> Maybe the code that resets "OBC" to "Pending" should also set "OB" to "Pending"?

This seems worthwhile to me.